src/go/k8s/pkg/resources/statefulset_update.go

```diff
@@ -0,0 +1,310 @@
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+// Package resources contains reconciliation logic for redpanda.vectorized.io CRD
+package resources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const requeueDuration = time.Second * 10
+
+func (r *StatefulSetResource) updateStsImage(
+	ctx context.Context, sts *appsv1.StatefulSet,
+) error {
+	upgrading := r.pandaCluster.Status.Upgrading
+
+	rpContainer, err := findContainer(sts.Spec.Template.Spec.Containers, redpandaContainerName)
+	if err != nil {
+		return err
+	}
+
+	newImage := fmt.Sprintf("%s:%s", r.pandaCluster.Spec.Image, r.pandaCluster.Spec.Version)
+	if rpContainer.Image == newImage && !upgrading {
+		return nil
+	}
+
+	if rpContainer.Image != newImage {
+		r.logger.Info("Starting cluster image update", "cluster image", newImage, "container image", rpContainer.Image)
+
+		// Mark cluster as being upgraded.
+		if err := r.updateUpgradingStatus(ctx, true); err != nil {
+			return err
+		}
+	}
+
+	if upgrading {
+		r.logger.Info("Continuing cluster image update", "cluster image", newImage)
+	}
+
+	podSpec := &sts.Spec.Template.Spec
+	if err := r.modifyPodImage(podSpec, newImage); err != nil {
+		return err
+	}
+
+	if err := r.partitionUpdateImage(ctx, sts, newImage); err != nil {
+		return err
+	}
+
+	// Update is complete for all pods (and all are ready). Set upgrading status to false.
+	if err := r.updateUpgradingStatus(ctx, false); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) updateUpgradingStatus(
+	ctx context.Context, upgrading bool,
+) error {
+	if !reflect.DeepEqual(upgrading, r.pandaCluster.Status.Upgrading) {
+		r.pandaCluster.Status.Upgrading = upgrading
+		if err := r.Status().Update(ctx, r.pandaCluster); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) partitionUpdateImage(
+	ctx context.Context, sts *appsv1.StatefulSet, newImage string,
+) error {
+	replicas := *sts.Spec.Replicas
+
+	// When a StatefulSet's partition number is set to `i`, only Pods with ordinal
+	// greater than or equal to `i` will be updated.
+	// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#partitions
+	for ordinal := replicas - 1; ordinal >= 0; ordinal-- {
+		// Update() on statefulset has not been called yet in this run, however,
+		// this could be a retry call in which case we skip the current partition.
+		poderr := r.podImageIdenticalToClusterImage(ctx, sts, newImage, ordinal)
+		if poderr == nil {
+			r.logger.Info("Pod already updated, skip", "ordinal", ordinal)
```

Comment:
so what if the pod is updated but it did not join the group and we retriggered reconcile on line 113, that's incorrect, right?

Reply:
Please help me out understand this. Here is a list of invariants:

- Each iteration "group-verifies" the _previously_ updated pod
- Each iteration updates 0 or 1 pod
- A pod is updated only if the _previous_ pod is "group-ready".

Note that `r.ensureRedpandaGroupsReady(sts, replicas, ordinal+1)` checks `ordinal+1` not `ordinal`.

Reply:
Oh you're right, then I guess we're fine. I missed that part, perhaps we could make it a bit more obvious but that could be refactoring after. Thanks for the explanation